### PR TITLE
gogen: Add type safe helper functions

### PR DESCRIPTION
**Pull request by lime008**
Add `New{Type}Publisher`, `New{Type}Subscriber`, `New{Type}Service` and `New{Type}Client` methods for the generated
message bindings to provide type safe interaction with publishers,
subscribers, services and clients.

### CODE REVIEW

**Review by lassilaiho — CHANGES_REQUESTED** [review]
The handler passed to generated subscriptions does not receive the generated subscription type. There should be a generated handler type as well.

File: pkg/gogen/templates.go [hunk 1]
```diff
@@ -112,6 +114,47 @@ func (t *{{$Md.Name}}) SetDefaults() {
...
+
+// New{{$Md.Name}}Subscription creates and returns a new subscription for the
+// {{$Md.Name}}
+func New{{$Md.Name}}Subscription(node *rclgo.Node, topic_name string, subscriptionCallback rclgo.SubscriptionCallback) (*{{$Md.Name}}Subscription, error) {
+	sub, err := node.NewSubscription(topic_name, {{$Md.Name}}TypeSupport, subscriptionCallback)
+	if err != nil {
+		return nil, err
+	}
+	return &{{$Md.Name}}Subscription{sub}, nil
+}
```
Comment (ezzkoram) [inline pkg/gogen/templates.go:151]: Would be nice to be able to just use API as this:
```
NewFooSubscription(node, "topic", func(msg Foo, info rclgo.RmwMessageInfo) {
    fmt.Printf("got message: %+v", msg)
})
```

```suggestion

type {{$Md.Name}}SubscriptionCallback func(msg {{$Md.Name}}, info rclgo.RmwMessageInfo)

// New{{$Md.Name}}Subscription creates and returns a new subscription for the
// {{$Md.Name}}
func New{{$Md.Name}}Subscription(node *rclgo.Node, topic_name string, subscriptionCallback {{$Md.Name}}SubscriptionCallback) (*{{$Md.Name}}Subscription, error) {

	callback := func(s *rclgo.Subscription) {
		var msg {{$Md.Name}}
		err, info := s.TakeMessage(msg)
		// TODO: handle error?
		subscriptionCallback(msg, info)
	}

	sub, err := node.NewSubscription(topic_name, {{$Md.Name}}TypeSupport, callback)
	if err != nil {
		return nil, err
	}
	return &{{$Md.Name}}Subscription{sub}, nil
}
```

Comment (lassilaiho) [reply pkg/gogen/templates.go:151]: The error could be passed to the handler to let the user decide how to handle the error.

```suggestion
type {{$Md.Name}}SubscriptionCallback func(msg *{{$Md.Name}}, info *rclgo.RmwMessageInfo, err error)

// New{{$Md.Name}}Subscription creates and returns a new subscription for the
// {{$Md.Name}}
func New{{$Md.Name}}Subscription(node *rclgo.Node, topic_name string, subscriptionCallback {{$Md.Name}}SubscriptionCallback) (*{{$Md.Name}}Subscription, error) {
	callback := func(s *rclgo.Subscription) {
		var msg {{$Md.Name}}
		info, err := s.TakeMessage(&msg)
		subscriptionCallback(&msg, info, err)
	}
	sub, err := node.NewSubscription(topic_name, {{$Md.Name}}TypeSupport, callback)
	if err != nil {
		return nil, err
	}
	return &{{$Md.Name}}Subscription{sub}, nil
}
```

Comment (lime008) [reply pkg/gogen/templates.go]: `{type}SubscriptionCallback` function type added and implemented according to the suggestion by @lassilaiho in commit: 3c63ebe.

**Review by lassilaiho — APPROVED** [review]

**Pull request merged**